src/Features/Core/Portable/DocumentSymbols/AbstractDocumentSymbolsService.cs

```diff
@@ -0,0 +1,185 @@
+                while (nodesToVisit.Count > startingCount)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var node = nodesToVisit.Pop();
+
+                    var symbol = GetSymbol(semanticModel, node, cancellationToken);
+
+                    if (symbol is { IsImplicitlyDeclared: false })
+                    {
+                        ImmutableArray<DocumentSymbolInfo> childrenSymbols;
+                        if (ShouldSkipSyntaxChildren(node, options))
+                        {
+                            childrenSymbols = ImmutableArray<DocumentSymbolInfo>.Empty;
+                        }
+                        else if (!ConsiderNestedNodesChildren(symbol))
+                        {
+                            childrenSymbols = ImmutableArray<DocumentSymbolInfo>.Empty;
+                            AddChildrenToVisit(nodesToVisit, node);
+                        }
+                        else
+                        {
+                            var currentCount = nodesToVisit.Count;
+                            AddChildrenToVisit(nodesToVisit, node);
+
+                            childrenSymbols = GetMembersInNode(semanticModel, tree, nodesToVisit, currentCount, options, cancellationToken);
+                            Debug.Assert(nodesToVisit.Count == currentCount);
+                        }
+
+                        memberBuilder.Add(CreateInfo(symbol, tree, declarationService, childrenSymbols, cancellationToken));
+                    }
+                    else
+                    {
+                        if (ShouldSkipSyntaxChildren(node, options))
+                        {
+                            continue;
+                        }
+                        AddChildrenToVisit(nodesToVisit, node);
+                    }
+                }
+
+                return memberBuilder.ToImmutable();
+
+                static void AddChildrenToVisit(ArrayBuilder<SyntaxNode> nodesToVisit, SyntaxNode node)
+                {
+                    var childSyntaxElements = node.ChildNodesAndTokens();
+                    // Push onto the stack in reverse order to make sure we visit in file order
+                    for (var i = childSyntaxElements.Count - 1; i >= 0; i--)
+                    {
+                        if (childSyntaxElements[i].AsNode() is { } child)
+                        {
+                            nodesToVisit.Push(child);
+                        }
+                    }
+                }
+            }
+        }
+
+        protected static ImmutableArray<TextSpan> GetDeclaringSpans(ISymbol symbol, SyntaxTree tree)
+        {
+            foreach (var location in symbol.Locations)
+            {
+                if (tree.Equals(location.SourceTree))
+                {
+                    return ImmutableArray.Create(location.SourceSpan);
```

Comment:
what if teh symbol has multiple locaitons in the same tree?  this will just return the first in Locations. Is that desirable/intentional?  Regardless of answer, please doc.